File: rust/src/protocol/v763/packets/play/serverbound/window_click.rs
//! Packet `window_click` (play/serverbound). Hand-written.

use crate::codec::uuid_codec::{self as uuid_c, Uuid};
use crate::codec::{
    bitset, chat_component, identifier, nbt, position, slot, string_codec as string, varint,
    varlong, BytesReader, BytesWriter, Reader, Writer,
};
use crate::errors::ProtocolError;
use crate::protocol::v763::states::ConnectionState;
use crate::protocol::v763::ServerboundPacket;

pub const PACKET_ID: i32 = 0x0B;

#[derive(Debug, Clone, PartialEq, Default)]
pub struct ChangedSlot {
    pub slot_index: i16,
    pub item: Option<slot::SlotData>,
}

#[derive(Debug, Clone, PartialEq, Default)]
pub struct WindowClick {
    pub window_id: u8,
    pub state_id: i32,
    pub slot_index: i16,
    pub mouse_button: i8,
    pub mode: i32,
    pub changed_slots: Vec<ChangedSlot>,
    pub carried_item: Option<slot::SlotData>,
}

impl WindowClick {
    pub fn decode(reader: &mut BytesReader<'_>) -> Result<Self, ProtocolError> {
        let window_id = reader.read_exact(1)?[0];
        let state_id = varint::read(reader)?;
        let sb = reader.read_exact(2)?;
        let slot_index = i16::from_be_bytes([sb[0], sb[1]]);
        let mouse_button = reader.read_exact(1)?[0] as i8;
        let mode = varint::read(reader)?;
        let n = varint::read(reader)? as usize;
        let mut changed_slots = Vec::with_capacity(n);
        for _ in 0..n {
            let sb = reader.read_exact(2)?;
            let s = i16::from_be_bytes([sb[0], sb[1]]);
            let item = slot::read(reader)?;
            changed_slots.push(ChangedSlot {
                slot_index: s,
                item,
            });
        }
        let carried_item = slot::read(reader)?;
        Ok(Self {
            window_id,
            state_id,
            slot_index,
            mouse_button,
            mode,
            changed_slots,
            carried_item,
        })
    }
}

impl ServerboundPacket for WindowClick {
    fn state(&self) -> ConnectionState {
        ConnectionState::Play
    }
    fn packet_id(&self) -> i32 {
        PACKET_ID
    }
    fn encode(&self, writer: &mut BytesWriter) -> Result<(), ProtocolError> {
        writer.write_all(&[self.window_id])?;
        varint::write(self.state_id, writer)?;
        writer.write_all(&self.slot_index.to_be_bytes())?;
        writer.write_all(&[self.mouse_button as u8])?;
        varint::write(self.mode, writer)?;
        varint::write(self.changed_slots.len() as i32, writer)?;
        for c in &self.changed_slots {
            writer.write_all(&c.slot_index.to_be_bytes())?;
            slot::write(c.item.as_ref(), writer)?;
        }
        slot::write(self.carried_item.as_ref(), writer)?;
        Ok(())
    }
}

File: python/minecraft_bot/protocol/v763/packets/play/serverbound/window_click.py
"""Packet `window_click` (play/serverbound, id 0x0B).

Inventory click. ``mode`` codes: 0=normal, 1=shift-click, 2=hotbar swap,
3=middle-click, 4=drop, 5=drag, 6=double-click. ``changed_slots`` is a
list of (slot_index, new_item) tuples reflecting the client's
optimistic prediction.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass

from minecraft_bot.codec import Reader, Writer, slot, varint

PACKET_ID = 0x0B


@dataclass(frozen=True, slots=True)
class ChangedSlot:
    slot_index: int                 # i16
    item: slot.SlotData | None


@dataclass(frozen=True, slots=True)
class WindowClick:
    window_id: int                  # u8
    state_id: int                   # varint
    slot_index: int                 # i16
    mouse_button: int               # i8
    mode: int                       # varint
    changed_slots: tuple[ChangedSlot, ...]
    carried_item: slot.SlotData | None


def decode(reader: Reader) -> WindowClick:
    wid = reader.read(1)[0]
    state = varint.read(reader)
    si, = struct.unpack(">h", reader.read(2))
    mb, = struct.unpack(">b", reader.read(1))
    mode = varint.read(reader)
    n = varint.read(reader)
    changed: list[ChangedSlot] = []
    for _ in range(n):
        s, = struct.unpack(">h", reader.read(2))
        item = slot.read(reader)
        changed.append(ChangedSlot(slot_index=s, item=item))
    carried = slot.read(reader)
    return WindowClick(
        window_id=wid, state_id=state, slot_index=si,
        mouse_button=mb, mode=mode,
        changed_slots=tuple(changed), carried_item=carried,
    )


def encode(packet: WindowClick, writer: Writer) -> None:
    writer.write(bytes([packet.window_id & 0xFF]))
    varint.write(packet.state_id, writer)
    writer.write(struct.pack(">h", packet.slot_index))
    writer.write(struct.pack(">b", packet.mouse_button))
    varint.write(packet.mode, writer)
    varint.write(len(packet.changed_slots), writer)
    for cs in packet.changed_slots:
        writer.write(struct.pack(">h", cs.slot_index))
        slot.write(cs.item, writer)
    slot.write(packet.carried_item, writer)
